auth/authentication.go

```diff
@@ -17,119 +16,163 @@ type Database interface {
 	GetApplicationByToken(token string) (*model.Application, error)
 	GetClientByToken(token string) (*model.Client, error)
 	GetPluginConfByToken(token string) (*model.PluginConf, error)
+
 	GetUserByName(name string) (*model.User, error)
 	GetUserByID(id uint) (*model.User, error)
+	CreateUser(user *model.User) error
+}
+
+// AuthenticationProvider provides authentication methods
+type AuthenticationProvider interface {
+	Authenticate(req *http.Request) (user *model.User, err error)
 }
 
 // Auth is the provider for authentication middleware
 type Auth struct {
-	DB Database
+	authenticators map[string]AuthenticationProvider
```

Comment:
Shouldn't we use a list instead of a map here so the order in which external authenticators are used is always the same and configurable?

Reply:
Makes sense. Good idea :+1: